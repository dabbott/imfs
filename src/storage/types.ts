export type File = { type: 'file'; data: Uint8Array }

export type Directory = {
  type: 'directory'
  entries: Record<string, Entry>
}

export type Entry = File | Directory
