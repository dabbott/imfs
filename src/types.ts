export type File<Data> = { type: 'file'; data: Data }

export type Directory<Data> = {
  type: 'directory'
  entries: Record<string, Node<Data>>
}

export type Node<Data> = File<Data> | Directory<Data>
