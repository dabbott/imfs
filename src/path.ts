export const sep = '/'

function preserveLeadingAndTrailingSlash(original: string, updated: string) {
  if (original.startsWith('/')) {
    updated = '/' + updated
  }

  if (updated[updated.length - 1] !== '/' && original.endsWith('/')) {
    updated = updated + '/'
  }

  return updated
}

export function normalize(filename: string) {
  const components = filename.split(sep).filter((component) => !!component)

  let i = 0
  let length = components.length - 1

  while (i < length) {
    if (components[i] === '.') {
      components.splice(i, 1)
      length--
    } else if (components[i] === '..' && i !== 0) {
      components.splice(i - 1, 2)
      length -= 2
      i -= 1
    } else if (components[i] === '' && components[i + 1] === '') {
      // Consecutive "/"
      components.splice(i, 1)
      length--
    } else {
      i++
    }
  }

  return preserveLeadingAndTrailingSlash(filename, components.join(sep))
}

export function join(...components: string[]) {
  let result = normalize(components.join(sep))

  if (result === '') {
    result = '.'
  }

  return result
}

export function extname(filename: string) {
  const index = filename.lastIndexOf('.')
  return index !== -1 ? filename.slice(index) : filename
}

export function basename(filename: string, extname?: string) {
  if (extname && filename.endsWith(extname)) {
    filename = filename.slice(0, -extname.length)
  }

  const sepIndex = filename.lastIndexOf(sep)

  // If the seperator index is the last character, omit it
  if (sepIndex === filename.length - 1) {
    return filename.slice(0, -1)
  }

  return filename.slice(sepIndex + 1)
}

export function dirname(filename: string) {
  let base = basename(filename)
  let result = filename.slice(0, -(base.length + 1))

  if (result === '') {
    if (filename.startsWith(sep)) {
      result = sep
    } else {
      result = '.'
    }
  }

  return result
}
