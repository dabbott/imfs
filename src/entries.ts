import { Nodes } from '.'
import { join } from './path'
import { Entry, Node } from './types'

function getEntries<Data>(entry: Entry<Data>): Entry<Data>[] {
  const [pathname, node] = entry

  return Nodes.isDirectory(node)
    ? Object.entries(node.children).map(([key, value]) => [
        join(pathname, key),
        value,
      ])
    : []
}

function createEntry<Data>(pathname: string, node: Node<Data>): Entry<Data> {
  return [pathname, node]
}

export const Entries = {
  createEntry,
  getEntries,
}
