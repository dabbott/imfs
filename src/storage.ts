import { join, sep } from './path'

export type File = Uint8Array

export type Entry = File | Directory

export interface Directory {
  [key: string]: Entry
}

export function isDirectory(entry: Entry): entry is Directory {
  return typeof entry === 'object'
}

export function isFile(entry: Entry): entry is File {
  return entry instanceof Uint8Array
}

export function contains<T extends string>(
  directory: Directory,
  name: T
): directory is Directory & { [key in T]: Entry } {
  return name in directory
}

export function hasFile<T extends string>(
  directory: Directory,
  name: T
): directory is Directory & { [key in T]: Entry } {
  return contains(directory, name) && isFile(directory[name])
}

export function hasDirectory<T extends string>(
  directory: Directory,
  name: T
): directory is Directory & { [key in T]: Directory } {
  return contains(directory, name) && isDirectory(directory[name])
}

export function locate(
  directory: Directory,
  filepath: string
):
  | {
      parent: Directory
      name: string
    }
  | undefined {
  const components = filepath.split(sep).filter((x) => x !== '')

  if (components.length === 0) return

  let current: Directory = directory

  for (let i = 0; i < components.length - 1; i++) {
    const component = components[i]

    if (hasDirectory(current, component)) {
      current = current[component]
      continue
    } else {
      const next: Directory = {}
      current[component] = next
      current = next
    }
  }

  return {
    parent: current,
    name: components[components.length - 1],
  }
}

export function exists(directory: Directory, filepath: string) {
  return !!locate(directory, filepath)
}

export function getDirectory(
  directory: Directory,
  filepath: string
): Directory | undefined {
  const target = locate(directory, filepath)

  if (!target) return

  if (!hasDirectory(target.parent, target.name)) return

  return target.parent[target.name]
}

export function getFile(
  directory: Directory,
  filepath: string
): Entry | undefined {
  const target = locate(directory, filepath)

  if (!target) return

  if (!hasFile(target.parent, target.name)) return

  return target.parent[target.name]
}

export function makeDirectory(directory: Directory, filepath: string) {
  const target = locate(directory, filepath)

  if (target) {
    target.parent[target.name] = {}
  } else {
    throw new Error(`Failed to locate ${filepath}`)
  }
}

export function readDirectory(
  directory: Directory,
  filepath: string
): string[] {
  const target = getDirectory(directory, filepath)

  if (!target) return []

  return Object.keys(target)
}

export function writeFile(
  directory: Directory,
  filepath: string,
  data: Uint8Array
) {
  const target = locate(directory, filepath)

  if (target) {
    target.parent[target.name] = data
  } else {
    throw new Error(`Failed to locate ${filepath}`)
  }
}

export function getPaths(directory: Directory): string[] {
  function inner(path: string, directory: Directory): string[] {
    const names = Object.keys(directory)

    return names
      .map((name) => join(path, name))
      .concat(
        ...names.map((name) => {
          if (hasDirectory(directory, name)) {
            return inner(join(path, name), directory[name])
          } else {
            return []
          }
        })
      )
  }

  return inner('', directory)
}

export function create(): Directory {
  return {}
}
