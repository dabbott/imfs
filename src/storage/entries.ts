import { withOptions } from 'tree-visit'
import { join } from '../path'
import { Entry, File, Directory } from './types'

function createFile(data: Uint8Array): File {
  return { type: 'file', data }
}

function createDirectory(entries: Directory['entries']): Directory {
  return { type: 'directory', entries }
}

function isFile(entry: Entry): entry is File {
  return entry.type === 'file'
}

function isDirectory(entry: Entry): entry is Directory {
  return entry.type === 'directory'
}

function readDirectory(directory: Directory) {
  return Object.keys(directory.entries)
}

function getEntry(directory: Directory, name: string): Entry | undefined {
  return directory.entries[name]
}

function hasEntry<T extends string>(
  directory: Directory,
  name: T
): directory is Directory & { value: { [key in T]: Entry } } {
  return name in directory.entries
}

const Traverse = withOptions<[string, Entry]>({
  getChildren: ([pathname, entry]) =>
    entry.type === 'directory'
      ? Object.entries(entry.entries).map(([key, value]) => [
          join(pathname, key),
          value,
        ])
      : [],
})

export const Entries = {
  createFile,
  createDirectory,
  isFile,
  isDirectory,
  readDirectory,
  getEntry,
  hasEntry,
  Traverse,
}
