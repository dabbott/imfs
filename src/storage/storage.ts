import produce from 'immer'
import { basename, dirname, join, normalize, sep } from '../path'
import { Entries } from './entries'
import { Entry } from './types'

type PathLike = string | string[]

function getPathComponents(filepath: string): string[] {
  return getPathComponentsNormalized(normalize(filepath))

  function getPathComponentsNormalized(filepath: string): string[] {
    if (filepath.startsWith('..')) {
      throw new Error(`Invalid path ${filepath}, can't go up past root.`)
    }

    if (filepath.startsWith('.') || filepath.startsWith('/')) {
      return getPathComponentsNormalized(filepath.slice(1))
    }

    if (filepath === '') return []

    return filepath.split(sep)
  }
}

const getComponentsInternal = (pathlike: PathLike) => {
  return typeof pathlike === 'string' ? getPathComponents(pathlike) : pathlike
}

function getEntry(root: Entry, pathlike: PathLike): Entry {
  const components = getComponentsInternal(pathlike)

  let i = 0
  let current: Entry = root

  while (i < components.length) {
    let component = components[i]

    if (current.type !== 'directory') {
      throw new Error(
        `File ${join(...components.slice(0, i))} is not a directory`
      )
    }

    const entry = Entries.getEntry(current, component)

    if (!entry) {
      throw new Error(`File ${join(...components)} not found`)
    }

    current = entry

    i++
  }

  return current
}

function readFile(root: Entry, pathlike: PathLike): Uint8Array {
  const components = getComponentsInternal(pathlike)
  const entry = getEntry(root, components)

  if (!Entries.isFile(entry)) {
    throw new Error(`Can't read, ${join(...components)} not a file`)
  }

  return entry.data
}

function readDirectory(root: Entry, pathlike: PathLike): string[] {
  const components = getComponentsInternal(pathlike)
  const entry = getEntry(root, components)

  if (!Entries.isDirectory(entry)) {
    throw new Error(`Can't read, ${join(...components)} not a directory`)
  }

  return Entries.readDirectory(entry)
}

function makeDirectory<T extends Entry>(root: T, path: string): T {
  const parentName = dirname(path)
  const newName = basename(path)

  return produce(root, (draft) => {
    const parent = getEntry(draft, parentName)

    if (!Entries.isDirectory(parent)) {
      throw new Error(
        `Can't create directory ${newName}, ${parentName} not a directory`
      )
    }

    parent.entries[newName] = Entries.createDirectory()
  })
}

function writeFile<T extends Entry>(
  root: T,
  path: string,
  data: Uint8Array
): T {
  const parentName = dirname(path)
  const newName = basename(path)

  return produce(root, (draft) => {
    const parent = getEntry(draft, parentName)

    if (!Entries.isDirectory(parent)) {
      throw new Error(
        `Can't create file ${newName}, ${parentName} not a directory`
      )
    }

    parent.entries[newName] = Entries.createFile(data)
  })
}

export const Storage = {
  getPathComponents,
  getEntry,
  readFile,
  readDirectory,
  makeDirectory,
  writeFile,
}
