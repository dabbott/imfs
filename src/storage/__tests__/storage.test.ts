import { basename } from '../../path'
import { Entries } from '../entries'
import { Storage } from '../storage'
import { Entry } from '../types'

function diagram<T>(root: Entry<T>) {
  return Entries.traversal().diagram(
    ['/', root],
    ([pathname, entry]) =>
      `${pathname === '/' ? '/' : basename(pathname)} (${entry.type})`
  )
}

it('gets path components', () => {
  expect(Storage.getPathComponents('/')).toEqual([])
  expect(Storage.getPathComponents('.')).toEqual([])
  expect(Storage.getPathComponents('/a')).toEqual(['a'])
  expect(Storage.getPathComponents('/nested')).toEqual(['nested'])
  expect(Storage.getPathComponents('/nested/b')).toEqual(['nested', 'b'])
})

it('gets entry', () => {
  const a = Entries.createFile(new Uint8Array())
  const b = Entries.createFile(new Uint8Array())
  const nested = Entries.createDirectory({ b })
  const root = Entries.createDirectory({ a, nested })

  expect(Storage.getEntry(root, '/')).toEqual(root)
  expect(Storage.getEntry(root, '.')).toEqual(root)
  expect(Storage.getEntry(root, '/a')).toEqual(a)
  expect(Storage.getEntry(root, 'a')).toEqual(a)
  expect(Storage.getEntry(root, '/nested')).toEqual(nested)
  expect(Storage.getEntry(root, '/nested/b')).toEqual(b)
  expect(Storage.getEntry(root, ['nested', 'b'])).toEqual(b)

  expect(() => Storage.getEntry(root, '..')).toThrowError(
    "Invalid path .., can't go up past root"
  )
  expect(() => Storage.getEntry(root, '/c')).toThrowError('File c not found')
  expect(() => Storage.getEntry(root, '/nested/c')).toThrowError(
    'File nested/c not found'
  )
  expect(() => Storage.getEntry(root, '/nested/b/c')).toThrowError(
    'File nested/b is not a directory'
  )
})

it('reading', () => {
  const a = Entries.createFile(new Uint8Array())
  const b = Entries.createFile(new Uint8Array())
  const nested = Entries.createDirectory({ b })
  const root = Entries.createDirectory({ a, nested })

  expect(Storage.readDirectory(root, '/')).toEqual(['a', 'nested'])
  expect(Storage.readDirectory(root, '/nested')).toEqual(['b'])
  expect(Storage.readFile(root, '/nested/b')).toEqual(b.data)
})

it('writing', () => {
  const root = Entries.createDirectory<string>()
  const withA = Storage.makeDirectory(root, '/a')
  const withAB = Storage.makeDirectory(withA, '/b')
  const withAC = Storage.writeFile(withA, '/c', 'hello')

  expect(diagram(root)).toMatchSnapshot()
  expect(diagram(withA)).toMatchSnapshot()
  expect(diagram(withAB)).toMatchSnapshot()
  expect(diagram(withAC)).toMatchSnapshot()
})
