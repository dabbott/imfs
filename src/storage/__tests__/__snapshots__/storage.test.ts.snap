// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`writing 1`] = `"/ (directory)"`;

exports[`writing 2`] = `
"/ (directory)
└── a (directory)"
`;

exports[`writing 3`] = `
"/ (directory)
├── a (directory)
└── b (directory)"
`;

exports[`writing 4`] = `
"/ (directory)
├── a (directory)
└── c (file)"
`;
